### PR TITLE
use SEEK_HOLE/SEEK_DATA for sparse copy

### PR DESCRIPTION
**Pull request by xinxinsh**
Signed-off-by: xinxin shu <xinxin.shu@intel.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
:+1: with a qa run

**Comment by yuyuyu101** [comment]
You also need to modify FileStoreBackend::_copy_range

**Comment by xinxinsh** [comment]
@yuyuyu101 , updated, help review

File: src/os/FileStore.cc [hunk 1]
```diff
@@ -3355,59 +3355,30 @@ int FileStore::_do_sparse_copy_range(int from, int to, uint64_t srcoff, uint64_t
 {
   dout(20) << __func__ << " " << srcoff << "~" << len << " to " << dstoff << dendl;
   int r = 0;
-  struct fiemap *fiemap = NULL;
-
+  map<uint64_t, uint64_t> exomap;
+  int64_t written = 0;
```
Comment (tchaikov) [inline src/os/FileStore.cc]: would be ideal if this definition (the variable of "written") could be move close to where it is first used, i.e. https://github.com/ceph/ceph/pull/5312/files#diff-0d2d3deefdf5127945a40ed3e04c154fR3372 .

**Comment by xinxinsh** [comment]
@tchaikov , updated

**Comment by yuyuyu101** [comment]
What if implements a new seek_data/seek_hole function instead of replacing?

**Comment by xinxinsh** [comment]
@yuyuyu101 , what do you mean "implement a new seek_data/seek_hole function", can you clarify a bit

**Comment by yuyuyu101** [comment]
@xinxinsh never mind, I'm looking at wrong :-)

Reviewed-by:

**Comment by liewegas** [comment]
:+1:, needs qa

**Comment by tchaikov** [comment]
tested at http://pulpito.ceph.com/ubuntu-2015-08-03_01:52:05-rados-wip-k3fu-testing---basic-multi/ .

the failed tests are due to http://tracker.ceph.com/issues/3058 , which is an xfs issue . and i checked the valgrind failure of http://pulpito.ceph.com/ubuntu-2015-08-03_01:52:05-rados-wip-k3fu-testing---basic-multi/996241 : not relevant .

**Pull request merged**